### PR TITLE
Support virtual bucketing on $path for hive unbucketed table

### PR DESCRIPTION
**Pull request by jessesleeping**
For unbucketed table, providing an option to virtually bucketing them
give us the ability to have addressable splits. This will be useful
when we introduce partial failure recovery in the future. User can
use hive connector session property "virtual_bucket_count" to control
this behavior.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java [hunk 1]
```diff
@@ -68,6 +70,11 @@
 
     private HiveBucketing() {}
 
+    public static int getVirtualBucket(int bucketCount, Path path)
+    {
+        return (hashBytes(0, wrappedBuffer(path.toString().getBytes())) & Integer.MAX_VALUE) % bucketCount;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: Never use `getBytes()` since that uses the platform character set. Instead, use `getBytes(UTF_8)`. However, `Slices` has a shortcut method `utf8Slice()` that takes a `String`.

But taking a step back, why not just do `path.hashCode()`?

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 2]
```diff
@@ -393,6 +401,25 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
                 .iterator();
     }
 
+    private List<InternalHiveSplit> getVirtuallyBucketedSplits(Path path, FileSystem fileSystem, InternalHiveSplitFactory splitFactory, int bucketCount)
+    {
+        // List all files in the partition and assign virtual bucket number to each of them
+        List<InternalHiveSplit> splitList = new ArrayList<>();
+        try {
+            new HiveFileIterator(path, fileSystem, directoryLister, namenodeStats, FAIL).forEachRemaining(file -> splitFactory.createInternalHiveSplit(file, getVirtualBucket(bucketCount, file.getPath()))
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Or write this as a stream
```java
return stream(new HiveFileIterator(path, fileSystem, directoryLister, namenodeStats, FAIL))
        .map(file -> splitFactory.createInternalHiveSplit(file, getVirtualBucket(bucketCount, file.getPath()))
        .filter(Optional::isPresent)
        .map(Optional::get)
        .collect(toImmutableList());

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 3]
```diff
@@ -429,15 +456,15 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
 
         // convert files internal splits
         List<InternalHiveSplit> splitList = new ArrayList<>();
-        for (int bucketNumber = 0; bucketNumber < Math.max(readBucketCount, partitionBucketCount); bucketNumber++) {
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: This static import change seems unrelated. It's fine to do, but please put it in a separate commit.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 4]
```diff
@@ -393,6 +401,25 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
...
+    private List<InternalHiveSplit> getVirtuallyBucketedSplits(Path path, FileSystem fileSystem, InternalHiveSplitFactory splitFactory, int bucketCount)
+    {
+        // List all files in the partition and assign virtual bucket number to each of them
+        List<InternalHiveSplit> splitList = new ArrayList<>();
+        try {
+            new HiveFileIterator(path, fileSystem, directoryLister, namenodeStats, FAIL).forEachRemaining(file -> splitFactory.createInternalHiveSplit(file, getVirtualBucket(bucketCount, file.getPath()))
+                    .ifPresent(splitList::add));
+        }
+        catch (NestedDirectoryNotAllowedException e) {
+            // Fail here to be on the safe side. This seems to be the same as what Hive does
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: This comment seems incorrect, since Hive doesn't have this virtual bucket behavior. This should probably be `NOT_SUPPORTED` or a new `HIVE_SUBDIRECTORY_NOT_SUPPORTED` with a message like
```java
"Hive table '%s' partition '%s' contains sub-subdirectory, which is not supported for virtual bucketing"
```

Comment (electrum) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Actually, I don't see why we need to fail here, since sub-directories will work fine with virtual bucketing. We should use `createInternalHiveSplitIterator()` which will have the normal behavior for directory walking.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java [hunk 5]
```diff
@@ -352,6 +355,11 @@ private void invokeNoMoreSplitsIfNecessary()
 
         // Bucketed partitions are fully loaded immediately since all files must be loaded to determine the file to bucket mapping
         if (tableBucketInfo.isPresent()) {
+            if (isVirtuallyBucketed(tableBucketInfo.get().getBucketColumns())) {
+                // For virtual bucket, bucket conversion must not be present because there is no physical partition bucket count
+                checkState(!bucketConversion.isPresent());
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Add a message here so we know what happened if this ever throws

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java (hunk 1 above)
Comment (jessesleeping) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveBucketing.java]: By using the same hash function in hive bucketing, virtual bucketed table will have the same behavior (e.g. hash distribution) as table that are bucketed on a single `VARCHAR` column. Although I am not sure what the actual benefit keeping them align but I think it's good to be consistent if the extra cost is not that significant.

**Comment by sopel39** [comment]
How does this relates to: https://github.com/prestodb/presto/issues/11876?

What is the advantage of bucketing by `$path` as it has very limited use case within queries to utilize bucket-by-bucket execution?

**Comment by jessesleeping** [comment]
@sopel39 I didn't know about #11876 before but I think they are not really targeting on the same problem. 

This PR is to provide a way for Hive unbucketed table to have addressable splits. For bucketed tables, a bucket can be considered as an execution unit as well as a failure recovery unit (although we don't have partial recovery yet). For unbucketed table, we need a similar way to group splits and identify these groups when needed. Whatever mapping that's fixed will do the work but "$path" seems more natural for Hive (bucketing is essentially based on order of file paths under the partition directory). We don't expect any query will benefit from virtual bucketing from a performance or resource usage perspective. 

I think #11876 is aiming at providing grouped execution across partitions and reduces memory footprint.

File: presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java (hunk 4 above)
Comment (jessesleeping) [reply presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java]: Yea, we should let the file iterator recursively walk down the partition directory with nested directory policy set to `RECURSE` instead of `FAIL`. Will fix it.

**Comment by sopel39** [comment]
> For bucketed tables, a bucket can be considered as an execution unit as well as a failure recovery unit (although we don't have partial recovery yet). For unbucketed table, we need a similar way to group splits and identify these groups when needed.

So the idea is to use `$path` virtual bucketing to provide failure-recovery unit?
But in order to really utilize such virtual bucket, queries have to be aggregated or joined on the `$path` column. If that is not the case, then the query will just execute as a non-bucketed query and the execution unit would be entire table. Please correct me if I'm wrong.

**Comment by wenleix** [comment]
@sopel39 : You are right, virtual buckets doesn't help failure recovery with join/aggregations.

It serves as two purpose:
- It make ScanFilterProject only query from un-bucketed table failure recoverable -- each "bucket" corresponding to a set of files.
- It opens opportunities for recoverable shuffle, think about how MapReduce works. Of course, there are other alternatives to recoverable shuffle, such as existing in-memory shuffle + checkpoint.

 
In general, we just need some deterministic way of partitioning the input table. Computing hash over $path is one way.

**Comment by sopel39** [comment]
> It make ScanFilterProject only query from un-bucketed table failure recoverable -- each "bucket" corresponding to a set of files.

Is there such mechanism at the moment?
I don't think `PlanFragmenter.GroupedExecutionTagger` will trigger for `$path` bucket, will it?
Why one cannot just restart split processing if no data was produced? Do we need `$path` for it?

**Comment by shixuan-fan** [comment]
Superceded by https://github.com/prestodb/presto/pull/13008
